subsys/testsuite/ztest/include/ztest_test_new.h

```diff
@@ -0,0 +1,323 @@
+ * checks for fast failures and initialization.
+ *
+ * @param name The name of the suite to run.
+ * @return Negative value if the test suite never ran; otherwise, return the number of failures.
+ */
+int z_ztest_run_test_suite(const char *name);
+
+/**
+ * @defgroup ztest_test Ztest testing macros
+ * @ingroup ztest
+ *
+ * This module eases the testing process by providing helpful macros and other
+ * testing structures.
+ *
+ * @{
+ */
+
+/**
+ * @brief Fail the currently running test.
+ *
+ * This is the function called from failed assertions and the like. You
+ * probably don't need to call it yourself.
+ */
+void ztest_test_fail(void);
+
+/**
+ * @brief Pass the currently running test.
+ *
+ * Normally a test passes just by returning without an assertion failure.
+ * However, if the success case for your test involves a fatal fault,
+ * you can call this function from k_sys_fatal_error_handler to indicate that
+ * the test passed before aborting the thread.
+ */
+void ztest_test_pass(void);
+
+/**
+ * @brief Skip the current test.
+ *
+ */
+void ztest_test_skip(void);
+
+/**
+ * @brief Do nothing, successfully.
+ *
+ * Unit test / setup function / teardown function that does
+ * nothing, successfully. Can be used as a parameter to
+ * ztest_unit_test_setup_teardown().
+ */
+static inline void unit_test_noop(void)
+{
+}
+
+#define Z_ZTEST(suite, fn, t_options)                                                              \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(void);                                                            \
+	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
+		.test_suite_name = STRINGIFY(suite),                                               \
+		.name = STRINGIFY(fn),                                                             \
+		.test = (_##suite##_##fn##_wrapper),                                               \
+		.thread_options = t_options,                                                       \
+	};                                                                                         \
+	static void _##suite##_##fn##_wrapper(void *data)                                          \
+	{                                                                                          \
+		ARG_UNUSED(data);                                                                  \
+		suite##_##fn();                                                                    \
+	}                                                                                          \
+	static inline void suite##_##fn(void)
+
+#define Z_ZTEST_F(suite, fn, t_options)                                                            \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(struct suite##_fixture *this);                                    \
```

Comment:
currently it is implicit to the user that `struct suite##_fixture` must be created. Would it be better to simply use `void *`? Given that setup function returns that.

Reply:
I personally prefer for this to be strongly typed, it really makes things nice when working with an IDE, but I can go either way on this.